### PR TITLE
Add capability of running snapshot builds.

### PR DESCRIPTION
**Pull request by codebrain**
Adds a build parameter "-snapshots" that internally switches everything over to the snapshots sources as opposed to staging sources.

Example command line: build.bat integrate es -snapshots

### CODE REVIEW

File: build/scripts/Products.fsx [hunk 1]
```diff
@@ -162,6 +175,14 @@ module Products =
                                 if directoryExists original |> not then
                                     Rename (InDir @@ (sprintf "kibana-%s" version.FullVersion)) (InDir @@ original)
                             | _ -> ()
+                        
+                        // Snapshots need renaming as folder inside zip is named differently
+                        let useSnapshots = getBuildParamOrDefault "snapshots" "$false"
+                        if (useSnapshots = "$true") then                                              
+                            let existing = InDir @@ (sprintf "%s-%s" product.Name (Snapshots.GetVersions |> Seq.head))
+                            let target = InDir @@ (sprintf "%s-%s" product.Name version.FullVersion)
+                            Rename target existing
```
[the remark this answers is not in the document]
Comment (codebrain) [reply build/scripts/Products.fsx]: Using an example:

The hosted zip filename is: 
`https://snapshots.elastic.co/7.0.0-alpha1-ea57ee52/downloads/elasticsearch/elasticsearch-7.0.0-alpha1-SNAPSHOT.zip`

This is downloaded locally into the `in` dir as: `elasticsearch-7.0.0-alpha1-ea57ee52.zip`

When extracted it creates a folder called: `elasticsearch-7.0.0-alpha1-snapshot`

This is then renamed to `elasticsearch-7.0.0-alpha1-ea57ee52`

When you re-run the build script with the same parameters (assuming it all downloaded and built successfully) then it will simply re-use the extracted folder, and if that is missing, will re-use the downloaded zip file.

In short, it caches as expected.

File: build/scripts/Snapshots.fsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+ServicePointManager.SecurityProtocol <- SecurityProtocolType.Ssl3 ||| SecurityProtocolType.Tls ||| SecurityProtocolType.Tls11 ||| SecurityProtocolType.Tls12;
+ServicePointManager.ServerCertificateValidationCallback <- (fun _ _ _ _ -> true)
+
+module Snapshots =
+
+    let GetVersions = (
+        use webClient = new System.Net.WebClient()
+        let url = "https://artifacts-api.elastic.co/v1/versions"
+        let versions = webClient.DownloadString url |> JsonValue.Parse
```
[the remark this answers is not in the document]
Comment (codebrain) [reply build/scripts/Snapshots.fsx]: At a later date...

File: build/scripts/Snapshots.fsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        let arrayValue = versions.GetProperty "versions"
+        arrayValue.AsArray()
+        |> Seq.rev
+        |> Seq.map (fun x -> x.AsString())
+     )
+
+    let GetSnapshotBuilds version = (
+       use webClient = new System.Net.WebClient()
+       let url = "https://artifacts-api.elastic.co/v1/versions/" + version + "/builds"
+       let versions = webClient.DownloadString url |> JsonValue.Parse
```
[the remark this answers is not in the document]
Comment (codebrain) [reply build/scripts/Snapshots.fsx]: At a later date...

File: build/scripts/Snapshots.fsx [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+       let arrayValue = versions.GetProperty "builds"
+       arrayValue.AsArray()
+       |> Seq.rev
+       |> Seq.map (fun x -> x.AsString())
+    )
+
+    let GetSnapshotBuildAssets product version build = (
+       use webClient = new System.Net.WebClient()
+       let url = "https://artifacts-api.elastic.co/v1/versions/" + version + "/builds/" + build
+       let versions = webClient.DownloadString url |> JsonValue.Parse
```
[the remark this answers is not in the document]
Comment (codebrain) [reply build/scripts/Snapshots.fsx]: At a later date...

File: build/scripts/Snapshots.fsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+
+ServicePointManager.SecurityProtocol <- SecurityProtocolType.Ssl3 ||| SecurityProtocolType.Tls ||| SecurityProtocolType.Tls11 ||| SecurityProtocolType.Tls12;
+ServicePointManager.ServerCertificateValidationCallback <- (fun _ _ _ _ -> true)
+
+module Snapshots =
+
+    let GetVersions = (
+        use webClient = new System.Net.WebClient()
+        let url = "https://artifacts-api.elastic.co/v1/versions"
```
[the remark this answers is not in the document]
Comment (codebrain) [reply build/scripts/Snapshots.fsx]: ++

File: build/scripts/Products.fsx [hunk 6]
```diff
@@ -96,14 +105,18 @@ module Products =
             | Compile ->
                 match product with
                 | Elasticsearch ->
-                    sprintf "%s/elasticsearch/elasticsearch-%s.zip" ArtifactDownloadsUrl version.FullVersion
+                    let useSnapshots = getBuildParamOrDefault "snapshots" "$false"
```
[the remark this answers is not in the document]
Comment (russcam) [reply build/scripts/Products.fsx]: `snapshots` value could be passed into `CreateFromProduct`, which would simplify the snapshot switching in different places in `ProductVersions`

File: build/scripts/Snapshots.fsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (russcam) [reply build/scripts/Snapshots.fsx]: IMO, it'd be much cleaner and simpler to implement now - would avoid successive `GetPropery` calls

**Comment by russcam** [comment]
scripts project is missing a reference to `Snapshots.fsx`

File: build/scripts/Snapshots.fsx [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+open FSharp.Data
+
+ServicePointManager.SecurityProtocol <- SecurityProtocolType.Ssl3 ||| SecurityProtocolType.Tls ||| SecurityProtocolType.Tls11 ||| SecurityProtocolType.Tls12;
+ServicePointManager.ServerCertificateValidationCallback <- (fun _ _ _ _ -> true)
+
+module Snapshots =
+
+    let private urlBase = "https://artifacts-api.elastic.co/v1/versions"
+
+    let GetVersions = (
```
Comment (russcam) [inline build/scripts/Snapshots.fsx]: should this be a function i.e.  `GetVersions () = ...` ?

File: build/scripts/Products.fsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (codebrain) [reply build/scripts/Products.fsx]: :+1:

**Comment by codebrain** [comment]
@Mpdreamz - you can now run:
`.\build.bat integrate es 6.5.0 skiptests -snapshots`
and get a 6.5.0 snapshot build :)
